pkg/resources/external_volume.go

```diff
@@ -0,0 +1,711 @@
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/internal/logging"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/internal/provider"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/schemas"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/customdiff"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+var externalVolumeSchema = map[string]*schema.Schema{
+	"name": {
+		Type:             schema.TypeString,
+		ForceNew:         true,
+		Required:         true,
+		Description:      blocklistedCharactersFieldDescription("Identifier for the external volume; must be unique for your account."),
+		DiffSuppressFunc: suppressIdentifierQuoting,
+	},
+	// A list is used as the order of storage locations matter. Storage location position in the list is used to select
+	// the active storage location - https://docs.snowflake.com/en/user-guide/tables-iceberg-storage#active-storage-location
+	// This is also why it has been left as one list with optional cloud dependent parameters, rather than splitting into
+	// one list per cloud provider.
+	"storage_location": {
+		Type:        schema.TypeList,
+		Required:    true,
+		MinItems:    1,
+		Description: "Set of named cloud storage locations in different regions and, optionally, cloud platforms. Minimum 1 required. Note that not all parameter combindations are valid as they depend on the given storage_provider. Consult [the docs](https://docs.snowflake.com/en/sql-reference/sql/create-external-volume#cloud-provider-parameters-cloudproviderparams) for more details on this.",
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"storage_location_name": {
+					Type:        schema.TypeString,
+					Required:    true,
+					Description: "Name of the storage location. Must be unique for the external volume.",
+				},
+				"storage_provider": {
+					Type:             schema.TypeString,
+					Required:         true,
+					ValidateDiagFunc: sdkValidation(sdk.ToStorageProvider),
+					DiffSuppressFunc: SuppressIfAny(NormalizeAndCompare(sdk.ToStorageProvider)),
+					Description:      fmt.Sprintf("Specifies the cloud storage provider that stores your data files. Valid values are (case-insensitive): %s.", possibleValuesListed(sdk.ValidStorageProviderString)),
+				},
+				"storage_base_url": {
+					Type:        schema.TypeString,
+					Required:    true,
+					Description: "Specifies the base URL for your cloud storage location.",
+				},
+				"storage_aws_role_arn": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					Description: "Specifies the case-sensitive Amazon Resource Name (ARN) of the AWS identity and access management (IAM) role that grants privileges on the S3 bucket containing your data files.",
+				},
+				"storage_aws_external_id": {
```

Comment:
I’ve followed the example of the [Storage Integration](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/pkg/resources/storage_integration.go#L64), which leaves this as a read only parameter. However, for both the Storage Integration and the External Volume I can see use cases in the future where users would want to set it. We should keep that in mind for the next iteration.